exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/OracleDialect.kt

```diff
@@ -195,6 +195,28 @@ internal object OracleFunctionProvider : FunctionProvider() {
         toString()
     }
 
+    override fun upsert(
+        table: Table,
+        data: List<Pair<Column<*>, Any?>>,
+        onUpdate: List<Pair<Column<*>, Expression<*>>>?,
+        where: Op<Boolean>?,
+        transaction: Transaction,
+        vararg keys: Column<*>
+    ): String {
+        val statement = super.upsert(table, data, onUpdate, where, transaction, *keys)
+
+        val dualTable = data.appendTo(QueryBuilder(true), prefix = "(SELECT ", postfix = " FROM DUAL) S") { (column, value) ->
+            registerArgument(column, value)
+            +" AS "
+            append(transaction.identity(column))
+        }.toString()
+
+        val (leftReserved, rightReserved) = " USING " to " ON "
+        val leftBoundary = statement.indexOf(leftReserved) + leftReserved.length
+        val rightBoundary = statement.indexOf(rightReserved)
+        return statement.replaceRange(leftBoundary, rightBoundary, dualTable)
```

Comment:
A technicality, but I'm not entirely sure about this choice. If anyone has a cleaner idea of how to replace this chunk in the middle of a statement, I'm open to alternatives. The other way I considered was:
```kt
return statement.replaceAfter(" USING ", dualTable) + statement.replaceBefore(" ON ", "")
```

Reply:
it's ok to make it this way